### PR TITLE
Improve the retry support for nondeterministic expressions

### PR DESCRIPTION
**Pull request by firestarman**
Contributes to https://github.com/NVIDIA/spark-rapids/issues/11649

This PR is trying to address some requirements described in issue #11649, but not all of them.

It introduces two new classes named `GpuExpressionRetryable` and `RetryStateTracker` to initially set up a fundamental support for the items 2 and 3 as below, and adds in the relevant unit tests.

>2. We need code changes so that if a retry happens on a non-deterministic expression that is outside of a checkpoint/restore, then we fail instead of retrying.
>3.  We also want a way to detect a non-deterministic expression being run outside of a checkpoint/restore retry block and throw an error from the plan so that when we can have tests validate that we have this covered.

And it also adds the integration tests for the function `rand()` being used in `HashAggregate`, `Generate`, `Projection`,  `ArrowEvalPython` and `Filter`. This is for the item 4, and it still does not cover all the cases where a nondeterministic expression can be used, but we are closer than before.

### CODE REVIEW

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
build

**Comment by firestarman** [comment]
The failure in CI is known issues, see https://github.com/NVIDIA/spark-rapids/issues/11790.

File: integration_tests/src/main/python/rand_test.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+@ignore_order(local=True)
+def test_filter_with_rand():
+    const_int_gen = IntegerGen(nullable=False, min_val=1, max_val=1, special_cases=[])
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: unary_op_df(spark, const_int_gen, num_slices=1).filter(f.rand(42) > 0.5)
+    )
+
+# See https://github.com/apache/spark/commit/9c0b803ba124a6e70762aec1e5559b0d66529f4d
+@ignore_order(local=True)
+@pytest.mark.skipif(is_before_spark_351(),
```
Comment (revans2) [inline integration_tests/src/main/python/rand_test.py:78]: Should there be a follow on issue for us to handle this properly?

Comment (firestarman) [reply integration_tests/src/main/python/rand_test.py:78]: Personally we don't need to do anything.  It is a pure Spark bug.
Spark 3.5.0 will complain the exception as below when evaluating a `rand()` in Generate **with the codegen disabled** and fail this test. 
```
E                   Caused by: java.lang.IllegalArgumentException: requirement failed: Nondeterministic expression org.apache.spark.sql.catalyst.expressions.Rand should be initialized before eval.
E                   	at scala.Predef$.require(Predef.scala:281)
E                   	at org.apache.spark.sql.catalyst.expressions.Nondeterministic.eval(Expression.scala:497)
E                   	at org.apache.spark.sql.catalyst.expressions.Nondeterministic.eval$(Expression.scala:495)
E                   	at org.apache.spark.sql.catalyst.expressions.RDG.eval(randomExpressions.scala:35)
E                   	at org.apache.spark.sql.catalyst.expressions.CreateArray.$anonfun$eval$1(complexTypeCreator.scala:95)
E                   	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
...
E                   	at org.apache.spark.sql.catalyst.expressions.CreateArray.eval(complexTypeCreator.scala:95)
E                   	at org.apache.spark.sql.catalyst.expressions.ExplodeBase.eval(generators.scala:375)
E                   	at org.apache.spark.sql.execution.GenerateExec.$anonfun$doExecute$8(GenerateExec.scala:108)
```
And Spark before 3.5.0 will throw the following exception and fail the test too.
```
E               pyspark.errors.exceptions.captured.AnalysisException: nondeterministic expressions are only allowed in Project, Filter, Aggregate or Window, found:
E               explode(array(rand(42L))),col
E               in operator Generate explode(array(rand(42))), false, [col#2].;
E               Project [col#2]
E               +- Generate explode(array(rand(42))), false, [col#2]
E                  +- LogicalRDD [a#0], false
```

While GPU supports Generate with `rand` for all the versions. So to make this test work we have to ignore it for Spark before 3.5.1.

**Comment by firestarman** [comment]
>I would also love to see some rand tests around

> join (we can restrict the range of rand to make it actually match)
> hash aggregate with some distinct operators so expand is called
> (Do we need/want rand in the key for an aggregate?)
> window operations

Is it ok to add them by another PR ?

**Comment by firestarman** [comment]
@revans2  Could you help review this again?  I'v addressed your comments. Thx in advance.

**Review by revans2 — APPROVED** [review]
Sorry I am still trying to get caught up on reviews. I am okay with finishing this in a follow on issue.

**Pull request merged**